File: src/ArmControl.cpp
#include <ArmControl.h>

ArmControl::ArmControl(const std::string &robot_name): name(robot_name)
{
    ros::NodeHandle nd;
    pathSub = nd.subscribe("/myPath", 1, &ArmControl::getPath, this);
    stateSub = nd.subscribe("/" + robot_name  + "/action_topic", 1, &ArmControl::notification_callback,this);
    positionSub = nd.subscribe("/my_gen3_lite/joint_states", 1, &ArmControl::getPosition, this);
    realPath.header.frame_id = "base_link";
    real_Path_pub=nd.advertise<nav_msgs::Path>("/realPath", 1);
    dh_params_list << 0, 0, 243.3 / 1000, 0,
                      M_PI / 2, 0, 10 / 1000, 0 + M_PI / 2,
                      M_PI, 280 / 1000, 0, 0 + M_PI / 2,
                      M_PI / 2, 0, 245 / 1000, 0 + M_PI / 2,
                      M_PI / 2, 0, 57 / 1000, 0,
                      -M_PI / 2, 0, 235 / 1000, 0 - M_PI / 2;
    success = ArmControl::setReferenceFrame();
}

void ArmControl::notification_callback(const kortex_driver::ActionNotification& notif)
{
    last_action_notification_event = notif.action_event;
    last_action_notification_id = notif.handle.identifier;
}

bool ArmControl::wait_for_action_end_or_abort()
{
    while (ros::ok())
    {
        if (last_action_notification_event.load() == kortex_driver::ActionEvent::ACTION_END)
        {
            ROS_INFO("Received ACTION_END notification for action %d", last_action_notification_id.load());
            return true;
        }
        else if (last_action_notification_event.load() == kortex_driver::ActionEvent::ACTION_ABORT)
        {
            ROS_INFO("Received ACTION_ABORT notification for action %d", last_action_notification_id.load());
            all_notifs_succeeded = false;
            return false;
        }
        ros::spinOnce();
    }
    return false;
}

bool ArmControl::goHome()
{
    ros::ServiceClient service_client_read_action = nd.serviceClient<kortex_driver::ReadAction>("/" + name + "/base/read_action");
    kortex_driver::ReadAction service_read_action;

    // The Home Action is used to home the robot. It cannot be deleted and is always ID #2:
    service_read_action.request.input.identifier = HOME_ACTION_IDENTIFIER;

    if (!service_client_read_action.call(service_read_action))
    {
        std::string error_string = "Failed to call ReadAction";
        ROS_ERROR("%s", error_string.c_str());
        return false;
    }

    // We can now execute the Action that we read 
    ros::ServiceClient service_client_execute_action = nd.serviceClient<kortex_driver::ExecuteAction>("/" + name + "/base/execute_action");
    kortex_driver::ExecuteAction service_execute_action;

    service_execute_action.request.input = service_read_action.response.output;
    
    if (service_client_execute_action.call(service_execute_action))
    {
        ROS_INFO("The Home position action was sent to the robot.");
    }
    else
    {
        std::string error_string = "Failed to call ExecuteAction";
        ROS_ERROR("%s", error_string.c_str());
        return false;
    }

    return wait_for_action_end_or_abort();
}

bool ArmControl::goPose(const geometry_msgs::Pose &pose, uint32_t id)
{
    kortex_driver::ConstrainedPose my_constrained_pose;
    kortex_driver::CartesianSpeed my_cartesian_speed;

    my_cartesian_speed.translation = 0.1f;
    my_cartesian_speed.orientation = 15.0f;

    my_constrained_pose.constraint.oneof_type.speed.push_back(my_cartesian_speed);

    my_constrained_pose.target_pose.x = pose.position.x;
    my_constrained_pose.target_pose.y = pose.position.y;
    my_constrained_pose.target_pose.z = pose.position.z;
    my_constrained_pose.target_pose.theta_x = 90.0;
    my_constrained_pose.target_pose.theta_y = 0.0;
    my_constrained_pose.target_pose.theta_z = 150.0;

    ros::ServiceClient service_client_execute_action = nd.serviceClient<kortex_driver::ExecuteAction>("/" + name + "/base/execute_action");
    kortex_driver::ExecuteAction service_execute_action;

    service_execute_action.request.input.oneof_action_parameters.reach_pose.push_back(my_constrained_pose);
    service_execute_action.request.input.name = "pose";
    service_execute_action.request.input.handle.identifier = id;
    service_execute_action.request.input.handle.action_type = kortex_driver::ActionType::REACH_POSE;

    last_action_notification_event = 0;
    if (service_client_execute_action.call(service_execute_action))
    {
    ROS_INFO("Pose was sent to the robot.");
    }
    else
    {
    std::string error_string = "Failed to call ExecuteAction on pose";
    ROS_ERROR("%s", error_string.c_str());
    return false;
    }

    // Waiting for the pose 1 to end
    wait_for_action_end_or_abort();
    
    return true;
}

bool ArmControl::setReferenceFrame()
{
    ros::ServiceClient service_client_set_cartesian_reference_frame = nd.serviceClient<kortex_driver::SetCartesianReferenceFrame>("/" + name + "/control_config/set_cartesian_reference_frame");
    kortex_driver::SetCartesianReferenceFrame service_set_cartesian_reference_frame;

    service_set_cartesian_reference_frame.request.input.reference_frame = kortex_driver::CartesianReferenceFrame::CARTESIAN_REFERENCE_FRAME_BASE;
    if (!service_client_set_cartesian_reference_frame.call(service_set_cartesian_reference_frame))
    {
        std::string error_string = "Failed to call SetCartesianReferenceFrame";
        ROS_ERROR("%s", error_string.c_str());
        return false;
    }
    else
    {
        ROS_INFO("Set reference to base frame.");
    }

    // Wait a bit
    std::this_thread::sleep_for(std::chrono::milliseconds(250));

    return true;
}

bool ArmControl::followPath()
{
    std::vector<double> current_joint_angles = {0, 0, 0, 0, 0, 0};
    for (int i = 0; i < myPath.poses.size(); i++)
    {
        // success=goPose(myPath.poses[i].pose, i);
        // if(!success)
        // {
        //     return false;
        // }
        std::vector<double> poses = {myPath.poses[i].pose.position.x, myPath.poses[i].pose.position.y, myPath.poses[i].pose.position.z};
        std::vector<double> target_angles = ArmControl::inverse_kinematics(poses, current_joint_angles);
        std::cout<<"target_angles: "<<target_angles[0]<<", "<<target_angles[1]<<", "<<target_angles[2]<<", "<<target_angles[3]<<", "<<target_angles[4]<<", "<<target_angles[5]<<std::endl;
        current_joint_angles = target_angles;
    }
    return true;
}

void ArmControl::getPath(const nav_msgs::Path::ConstPtr msg)
{
    myPath = *msg;
}

void ArmControl::getPosition(const sensor_msgs::JointState::ConstPtr msg)
{
    positions = msg->position;
    ArmControl::plotRealPath();
}

void ArmControl::plotRealPath()
{
    geometry_msgs::PoseStamped real_poses;
    real_poses.header.frame_id = "base_link";
    realPath.header.stamp = ros::Time::now();
    real_poses.header.stamp=ros::Time::now();
    std::vector<double> tip_pose = get_tip_pose_DH(positions);
    real_poses.pose.position.x = tip_pose[0];
    real_poses.pose.position.y = tip_pose[1];
    real_poses.pose.position.z = tip_pose[2];
    realPath.poses.push_back(real_poses);
    real_Path_pub.publish(realPath);
}

Eigen::Matrix4d ArmControl::gen_T(float theta, float a, float alpha, float d) {
    Eigen::Matrix4d T;
    T << std::cos(theta), -std::cos(alpha)*std::sin(theta), std::sin(theta)*std::sin(alpha), a*std::cos(theta),
         std::sin(theta), std::cos(alpha)*std::cos(theta), -std::cos(theta)*std::sin(alpha), a*std::sin(theta),
         0, std::sin(alpha), std::cos(alpha), d,
         0, 0, 0, 1;
    return T;
}

Eigen::Matrix4d ArmControl::forward_kinematics(std::vector<double> theta){
    theta[1] += M_PI / 2;
    theta[2] += M_PI / 2;
    theta[3] += M_PI / 2;
    theta[4] += M_PI;
    theta[5] += M_PI / 2;
    Eigen::VectorXd alpha(6);
    alpha << M_PI / 2, M_PI, M_PI / 2, M_PI / 2, M_PI / 2, 0;
    Eigen::VectorXd a(6);
    a << 0, 280, 0, 0, 0, 0;
    Eigen::VectorXd d(6);
    d << 128.3 + 115.0, 30, 20, 140 + 105, 28.5 + 28.5, 105 + 130;
    Eigen::Matrix4d T0_1 = ArmControl::gen_T(theta[0], a[0], alpha[0], d[0]);
    Eigen::Matrix4d T1_2 = ArmControl::gen_T(theta[1], a[1], alpha[1], d[1]);
    Eigen::Matrix4d T2_3 = ArmControl::gen_T(theta[2], a[2], alpha[2], d[2]);
    Eigen::Matrix4d T3_4 = ArmControl::gen_T(theta[3], a[3], alpha[3], d[3]);
    Eigen::Matrix4d T4_5 = ArmControl::gen_T(theta[4], a[4], alpha[4], d[4]);
    Eigen::Matrix4d T5_6 = ArmControl::gen_T(theta[5], a[5], alpha[5], d[5]);
    Eigen::Matrix4d T_rote;
    T_rote << 0, -1, 0, 0,
              1, 0, 0, 0,
              0, 0, 1, 0,
              0, 0, 0, 1;
    Eigen::Matrix4d T0_6 = T0_1 * T1_2 * T2_3 * T3_4 * T4_5 * T5_6 * T_rote;
    return T0_6;
}

std::vector<double> ArmControl::get_tip_pose_DH(std::vector<double> theta) {
    Eigen::Matrix4d T0_6 = ArmControl::forward_kinematics(theta);
    std::vector<double> tip_pose = {T0_6(0, 3)/1000, T0_6(1, 3)/1000, T0_6(2, 3)/1000};
    return tip_pose;
}

// 目标函数
double ArmControl::objective_function(std::vector<double> theta, std::vector<double> target_position) {
    Eigen::Vector3d target_position_vec = Eigen::Map<Eigen::Vector3d>(target_position.data(), target_position.size());
    Eigen::Vector3d end_effector_position = forward_kinematics(theta).block<3, 1>(0, 3);
    return (end_effector_position - target_position_vec).norm();
}

double ArmControl::constrain_angle(double angle) {
    return std::fmod(angle + M_PI, 2*M_PI) - M_PI;
}

// 最小化目标函数
std::vector<double> ArmControl::minimize(std::function<double(std::vector<double>)> func, std::vector<double> initial_guess) {
    /// 使用简单的梯度下降法作为示例，实际中可能需要更复杂的优化算法
    double alpha = 0.1;
    std::vector<double> theta = initial_guess;
    for (int i = 0; i < 1000; ++i) {
        std::vector<double> gradient(theta.size());
        for (int j = 0; j < theta.size(); ++j) {
            std::vector<double> theta_plus = theta;
            theta_plus[j] += 0.001;
            std::vector<double> theta_minus = theta;
            theta_minus[j] -= 0.001;
            gradient[j] = (func(theta_plus) - func(theta_minus)) / 0.002;
        }
        for (int j = 0; j < theta.size(); ++j) {
            theta[j] -= alpha * gradient[j];
            theta[j] = constrain_angle(theta[j]);
        }
    }
    return theta;
}

// 逆运动学解算函数
std::vector<double> ArmControl::inverse_kinematics(std::vector<double> target_position, std::vector<double> current_joint_angles) {
    std::function<double(std::vector<double>)> func = [this, target_position](std::vector<double> theta) {
        return objective_function(theta, target_position);
    };
    return minimize(func, current_joint_angles);
}

File: script/ArmControl.py
import sys
import rospy
import time
import math
from kortex_driver.srv import *
from kortex_driver.msg import *
from nav_msgs.msg import Path
from geometry_msgs.msg import Pose
from sensor_msgs.msg import JointState
from geometry_msgs.msg import PoseStamped
from std_msgs.msg import Header
import numpy as np
import os
sys.path.append(os.path.dirname(os.path.abspath(__file__)))
from Arm import Gen3LiteArm
from scipy.spatial.transform import Rotation

class ArmControl(object):
    def __init__(self):
        try:
            rospy.init_node('arm_control')

            self.HOME_ACTION_IDENTIFIER = 2

            self.action_topic_sub = None
            self.all_notifs_succeeded = True

            self.all_notifs_succeeded = True

            # Get node params
            self.robot_name = 'my_gen3_lite'

            rospy.loginfo("Using robot_name " + self.robot_name)

            # Init class members
            self.my_path_ = None
            self.curr_state_ = None
            self.plot = False
            self.real_path_ = Path()
            self.real_path_.header.frame_id = 'base_link'
            self.arm_model = Gen3LiteArm()
            self.target_pose_list = []

            # Init the subscribers and publishers
            self.path_subscriber = rospy.Subscriber("/myPath", Path, self.get_path, queue_size=1)
            self.curr_state_sub = rospy.Subscriber("/my_gen3_lite/joint_states", JointState, self.get_curr_state, queue_size=1)
            self.real_path_publisher = rospy.Publisher("/real_path", Path, queue_size=1)
            
            # Init the action topic subscriber
            self.action_topic_sub = rospy.Subscriber("/" + self.robot_name + "/action_topic", ActionNotification, self.cb_action_topic)
            self.last_action_notif_type = None

            # Init the services
            clear_faults_full_name = '/' + self.robot_name + '/base/clear_faults'
            rospy.wait_for_service(clear_faults_full_name)
            self.clear_faults = rospy.ServiceProxy(clear_faults_full_name, Base_ClearFaults)

            read_action_full_name = '/' + self.robot_name + '/base/read_action'
            rospy.wait_for_service(read_action_full_name)
            self.read_action = rospy.ServiceProxy(read_action_full_name, ReadAction)

            execute_action_full_name = '/' + self.robot_name + '/base/execute_action'
            rospy.wait_for_service(execute_action_full_name)
            self.execute_action = rospy.ServiceProxy(execute_action_full_name, ExecuteAction)

            set_cartesian_reference_frame_full_name = '/' + self.robot_name + '/control_config/set_cartesian_reference_frame'
            rospy.wait_for_service(set_cartesian_reference_frame_full_name)
            self.set_cartesian_reference_frame = rospy.ServiceProxy(set_cartesian_reference_frame_full_name, SetCartesianReferenceFrame)

            activate_publishing_of_action_notification_full_name = '/' + self.robot_name + '/base/activate_publishing_of_action_topic'
            rospy.wait_for_service(activate_publishing_of_action_notification_full_name)
            self.activate_publishing_of_action_notification = rospy.ServiceProxy(activate_publishing_of_action_notification_full_name, OnNotificationActionTopic)
        except:
            self.is_init_success = False
        else:
            self.is_init_success = True

        
    
    def cb_action_topic(self, notif):
            self.last_action_notif_type = notif.action_event

    def wait_for_action_end_or_abort(self):
        while not rospy.is_shutdown():
            if (self.last_action_notif_type == ActionEvent.ACTION_END):
                rospy.loginfo("Received ACTION_END notification")
                return True
            elif (self.last_action_notif_type == ActionEvent.ACTION_ABORT):
                rospy.loginfo("Received ACTION_ABORT notification")
                self.all_notifs_succeeded = False
                return False
            else:
                time.sleep(0.01)

    def example_clear_faults(self):
        try:
            self.clear_faults()
        except rospy.ServiceException:
            rospy.logerr("Failed to call ClearFaults")
            return False
        else:
            rospy.loginfo("Cleared the faults successfully")
            rospy.sleep(2.5)
            return True
    
    def go_home(self):
        # The Home Action is used to home the robot. It cannot be deleted and is always ID #2:
        req = ReadActionRequest()
        req.input.identifier = self.HOME_ACTION_IDENTIFIER
        self.last_action_notif_type = None
        try:
            res = self.read_action(req)
        except rospy.ServiceException:
            rospy.logerr("Failed to call ReadAction")
            return False
        # Execute the HOME action if we could read it
        else:
            # What we just read is the input of the ExecuteAction service
            req = ExecuteActionRequest()
            req.input = res.output
            rospy.loginfo("Sending the robot home...")
            try:
                self.execute_action(req)
            except rospy.ServiceException:
                rospy.logerr("Failed to call ExecuteAction")
                return False
            else:
                return self.wait_for_action_end_or_abort()
    
    def example_set_cartesian_reference_frame(self):
        # Prepare the request with the frame we want to set
        req = SetCartesianReferenceFrameRequest()
        req.input.reference_frame = CartesianReferenceFrame.CARTESIAN_REFERENCE_FRAME_MIXED

        # Call the service
        try:
            self.set_cartesian_reference_frame()
        except rospy.ServiceException:
            rospy.logerr("Failed to call SetCartesianReferenceFrame")
            return False
        else:
            rospy.loginfo("Set the cartesian reference frame successfully")
            return True
        
    def example_subscribe_to_a_robot_notification(self):
        # Activate the publishing of the ActionNotification
        req = OnNotificationActionTopicRequest()
        rospy.loginfo("Activating the action notifications...")
        try:
            self.activate_publishing_of_action_notification(req)
        except rospy.ServiceException:
            rospy.logerr("Failed to call OnNotificationActionTopic")
            return False
        else:
            rospy.loginfo("Successfully activated the Action Notifications!")

        rospy.sleep(1.0)
        return True
    
    def go_to_pose(self, pose, id):
        my_cartesian_speed = CartesianSpeed()
        my_cartesian_speed.translation = 0.2 # m/s
        my_cartesian_speed.orientation = 20  # deg/s

        my_constrained_pose = ConstrainedPose()
        my_constrained_pose.constraint.oneof_type.speed.append(my_cartesian_speed)

        my_constrained_pose.target_pose.x = pose[0]
        my_constrained_pose.target_pose.y = pose[1]
        my_constrained_pose.target_pose.z = pose[2]
        my_constrained_pose.target_pose.theta_x = pose[3]
        my_constrained_pose.target_pose.theta_y = pose[4]
        my_constrained_pose.target_pose.theta_z = pose[5]

        req = ExecuteActionRequest()
        req.input.oneof_action_parameters.reach_pose.append(my_constrained_pose)
        req.input.name = "pose"+str(id)
        req.input.handle.action_type = ActionType.REACH_POSE
        req.input.handle.identifier = 1001

        rospy.loginfo("Sending pose 1...")
        self.last_action_notif_type = None
        try:
            self.execute_action(req)
        except rospy.ServiceException:
            rospy.logerr("Failed to send" + req.input.name)
            return False
        else:
            rospy.loginfo("Waiting for"+req.input.name+"to finish...")

        return self.wait_for_action_end_or_abort()
    
    def get_path(self, msg):
        self.my_path_ = msg
    
    def follow_path(self):
        self.plot=True
        if self.my_path_ == None:
            rospy.logerr("No path received yet")
            return False
        
        # First calculate the target poses
        for i in range(len(self.my_path_.poses)):
            ref_ee_pose=[self.my_path_.poses[i].pose.position.x,self.my_path_.poses[i].pose.position.y,self.my_path_.poses[i].pose.position.z]
            ev_angles=self.arm_model.arm.inverse_kinematics(ref_ee_pose,np.zeros(6))
            ev_poses=self.forward_kinematics(ev_angles)
            self.target_pose_list.append(ev_poses)
        
        # Then go to each target pose
        for i in range(len(self.target_pose_list)):
            if not self.go_to_pose(self.target_pose_list[i],i):
                return False
        return True
    
    def get_curr_state(self, msg):
        self.curr_state_ = np.array(msg.position[0:6])
        # self.curr_state_ = msg
        self.plot_real_path()

    def plot_real_path(self):
        if self.plot == False:
            return
        # 创建一个 PoseStamped 消息
        real_poses = PoseStamped()
        real_poses.header.frame_id = "base_link"
        real_poses.header.stamp = rospy.Time.now()
        self.real_path_.header.stamp = rospy.Time.now()

        # 获取机械臂末端的实际位置
        tip_pose = self.forward_kinematics(self.curr_state_)

        # 填充 PoseStamped 消息
        real_poses.pose.position.x = tip_pose[0]
        real_poses.pose.position.y = tip_pose[1]
        real_poses.pose.position.z = tip_pose[2]

        # 将 PoseStamped 消息添加到 Path 中
        self.real_path_.poses.append(real_poses)

        # 发布实际路径
        self.real_path_publisher.publish(self.real_path_)

    def gen_T(self,theta,a,alpha,d):
        T=np.array([[np.cos(theta),-np.cos(alpha)*np.sin(theta),np.sin(theta)*np.sin(alpha),a*np.cos(theta)],
                    [np.sin(theta),np.cos(alpha)*np.cos(theta),np.cos(theta)*-np.sin(alpha),a*np.sin(theta)],
                    [0,np.sin(alpha),np.cos(alpha),d],
                    [0,0,0,1]])
        return T

    def get_transform_matrix(self,theta):
        theta[1]+=np.pi/2
        theta[2]+=np.pi/2
        theta[3]+=np.pi/2
        theta[4]+=np.pi
        theta[5]+=np.pi/2
        alpha=np.array([np.pi/2,np.pi,np.pi/2,np.pi/2,np.pi/2,0])
        a=np.array([0,280,0,0,0,0])
        d=np.array([128.3+115.0,30,20,140+105,28.5+28.5,105+130])
        T0_1=self.gen_T(theta[0],a[0],alpha[0],d[0])
        T1_2=self.gen_T(theta[1],a[1],alpha[1],d[1])
        T2_3=self.gen_T(theta[2],a[2],alpha[2],d[2])
        T3_4=self.gen_T(theta[3],a[3],alpha[3],d[3])
        T4_5=self.gen_T(theta[4],a[4],alpha[4],d[4])
        T5_6=self.gen_T(theta[5],a[5],alpha[5],d[5])
        T_rote=np.array([[0,-1,0,0],[1,0,0,0],[0,0,1,0],[0,0,0,1]])
        T0_6=T0_1@T1_2@T2_3@T3_4@T4_5@T5_6@T_rote
        return T0_6

    def forward_kinematics(self,state):
        # thetas=state.position[0:6]
        # thetas=np.array(thetas)
        T = self.get_transform_matrix(state)
        rotation_matrix = T[:3, :3]
        rotation = Rotation.from_matrix(rotation_matrix)
        euler_angles = rotation.as_euler('xyz', degrees=True)
        z=euler_angles[2]
        y=euler_angles[1]
        x=euler_angles[0]
        x_l=T[0,3]/1000
        y_l=T[1,3]/1000
        z_l=T[2,3]/1000
        return [x_l,y_l,z_l,x,y,z]
    
    def main(self):
        if not self.is_init_success:
            rospy.logerr("Initialization failed")
            return False
        
        if not self.example_clear_faults():
            return False
        
        if not self.example_set_cartesian_reference_frame():
            return False
        
        if not self.example_subscribe_to_a_robot_notification():
            return False
        
        if not self.go_home():
            return False
        

        while not rospy.is_shutdown():
            if self.my_path_ != None:
                break
        if not self.follow_path():
            return False
        rospy.loginfo("Complete painting!")
        return True

if __name__ == '__main__':
    arm_control = ArmControl()
    arm_control.main()